editor/src/messages/tool/tool_messages/path_tool.rs

```diff
@@ -502,135 +509,149 @@ impl PathToolData {
-			// Otherwise, select the first point within the threshold.
-			if !(already_selected && extend_selection) {
-				if let Some(updated_selection_info) = shape_editor.change_point_selection(
-					&document.network_interface,
-					input.mouse.position,
-					SELECTION_THRESHOLD,
-					extend_selection,
-					path_overlay_mode,
-					self.frontier_handles_info.clone(),
-				) {
-					selection_info = updated_selection_info;
-				}
+		let mouse_down_intent = {
+			// Check if the point is already selected; if not, select the first point within the threshold (in pixels)
+			// Don't select the points which are not shown currently in PathOverlayMode
+			if let Some((already_selected, selection_info)) = shape_editor.get_point_selection_state(
+				&document.network_interface,
+				input.mouse.position,
+				SELECTION_THRESHOLD,
+				path_overlay_mode,
+				self.frontier_handles_info.clone(),
+			) {
+				MouseDownIntent::UpdatePoint { already_selected, selection_info }
+			} else if lasso_select {
+				MouseDownIntent::Draw
+			} else if let Some(closed_segment) = &self.segment {
+				MouseDownIntent::MoldSegment { segment: closed_segment }
+			} else if let Some(layer) = document.click(input) {
+				MouseDownIntent::SelectLayer { layer }
+			} else {
+				MouseDownIntent::Draw
 			}
+		};
 
-			if let Some(selected_points) = selection_info {
-				self.drag_start_pos = input.mouse.position;
+		match mouse_down_intent {
+			MouseDownIntent::UpdatePoint { already_selected, mut selection_info } => {
+				responses.add(DocumentMessage::StartTransaction);
 
-				// If selected points contain only handles and there was some selection before, then it is stored and becomes restored upon release
-				let mut dragging_only_handles = true;
-				for point in &selected_points.points {
-					if matches!(point.point_id, ManipulatorPointId::Anchor(_)) {
-						dragging_only_handles = false;
-						break;
+				self.last_clicked_point_was_selected = already_selected;
+
+				// If the point is already selected and shift (`extend_selection`) is used, keep the selection unchanged.
+				// Otherwise, select the first point within the threshold.
+				if !(already_selected && extend_selection) {
+					if let Some(updated_selection_info) = shape_editor.change_point_selection(
+						&document.network_interface,
+						input.mouse.position,
+						SELECTION_THRESHOLD,
+						extend_selection,
+						path_overlay_mode,
+						self.frontier_handles_info.clone(),
+					) {
+						selection_info = updated_selection_info;
 					}
 				}
-				if dragging_only_handles && !self.handle_drag_toggle && !old_selection.is_empty() {
-					self.saved_points_before_handle_drag = old_selection;
-				}
 
-				if handle_drag_from_anchor {
-					if let Some((layer, point)) = shape_editor.find_nearest_point_indices(&document.network_interface, input.mouse.position, SELECTION_THRESHOLD) {
-						// Check that selected point is an anchor
```

Comment:
didn't change almost any of this logic. The diff just seems large because I moved the pieces from the if/else chain into a match statement